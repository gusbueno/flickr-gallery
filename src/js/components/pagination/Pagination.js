import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import styles from './Pagination.scss'

class Pagination extends PureComponent {
  _renderPaginationButtons () {
    const buttons = []
    const { page, pages } = this.props
    page > 1 && buttons.push(<div onClick={() => this.props.onNavigateToPage(page - 1)} key='back' className={styles['page-button']}>Back</div>)
    const toPage = (page + 10) <= pages ? page + 9 : pages
    for (let i = page; i <= toPage; i++) {
      buttons.push(<div onClick={() => this.props.onNavigateToPage(i)} key={i} className={[styles['page-button'], i === page && styles['current-page']].join(' ')}>{i}</div>)
    }
    page < pages && buttons.push(<div onClick={() => this.props.onNavigateToPage(page + 1)} key='next' className={styles['page-button']}>Next</div>)
    return buttons
  }

  render () {
    return (
      <div className={styles['pagination']}>
        {this.props.pages && this._renderPaginationButtons()}
      </div>
    )
  }
}

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
  pages: PropTypes.any,
  onNavigateToPage: PropTypes.func.isRequired
}

export default Pagination
