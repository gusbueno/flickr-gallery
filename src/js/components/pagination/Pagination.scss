.pagination {
  display: flex;
  flex: 1;
  margin-top: 10px;
  justify-content: center;
  
  .buttons-wrapper {
    width: 300px;
    height: 50px;
  }

  .page-button {
    padding: 10px;
    &:hover {
      cursor: pointer;
    }

    &.current-page {
      background-color: black;
      color: white;
    }
  }
}