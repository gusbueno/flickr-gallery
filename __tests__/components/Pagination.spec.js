'use strict'

/* global describe, it, expect, jest */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import Pagination from '../../src/js/components/pagination/Pagination'

describe('<Pagination /> Component', () => {
  it('should match with the snapshot', () => {
    const tree = renderer.create(
      <Pagination page={1} pages={20} onNavigateToPage={() => {}} />
    )
    const json = tree.toJSON()
    expect(json).toMatchSnapshot()
  })

  const onNavigateToPageMock = jest.fn()
  const paginationShallow = shallow(
    <Pagination page={1} pages={20} onNavigateToPage={onNavigateToPageMock} />
  )

  it('should exists', () => {
    expect(paginationShallow.length).toEqual(1)
  })

  it('should render pagination wrapper', () => {
    expect(paginationShallow.find('.pagination').length).toEqual(1)
  })

  it('should render eleven buttons', () => {
    expect(paginationShallow.find('.page-button').length).toEqual(11)
  })

  it('should render last button as Next button if current page is not the last', () => {
    expect(paginationShallow.find('.page-button').last().text()).toEqual('Next')
  })

  it('should render Back button if current page is not the first', () => {
    paginationShallow.setProps({ page: 2 })
    expect(paginationShallow.find('.page-button').first().text()).toEqual('Back')
  })

  it('should not render Next button if current page is the last', () => {
    paginationShallow.setProps({ page: 20 })
    expect(paginationShallow.find('.page-button').last().text()).not.toEqual('Next')
  })

  it('should call onNavigateToPage with page number on click some button', () => {
    paginationShallow.setProps({ page: 2 })
    paginationShallow.find('.page-button').at(3).simulate('click')
    expect(onNavigateToPageMock).toHaveBeenCalledWith(4)
  })

  it('should call onNavigateToPage with the before page number on click Back button', () => {
    paginationShallow.find('.page-button').first().simulate('click')
    expect(onNavigateToPageMock).toHaveBeenCalledWith(1)
  })

  it('should call onNavigateToPage with the next page number on click Next button', () => {
    paginationShallow.find('.page-button').last().simulate('click')
    expect(onNavigateToPageMock).toHaveBeenCalledWith(3)
  })
})
