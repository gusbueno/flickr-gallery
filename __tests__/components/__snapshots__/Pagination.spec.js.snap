// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Component should match with the snapshot 1`] = `
<div
  className="pagination"
>
  <div
    className="page-button current-page"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    4
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    6
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    7
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    8
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    9
  </div>
  <div
    className="page-button false"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="page-button"
    onClick={[Function]}
  >
    Next
  </div>
</div>
`;
